### PR TITLE
mimic: build/ops: Cython 0.29 removed support for subinterpreters: raises ImportError: Interpreter change detected ...

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/39592
http://tracker.ceph.com/issues/39593

### CODE REVIEW

**Comment by yuriw** [comment]
[wip-yuri-testing-2019-05-06-1633-mimic](https://trello.com/c/7ADkP0ea/607-wip-yuri-testing-2019-05-06-1633-mimic)

**Comment by yuriw** [comment]
@smithfarm this likely broke centos build:
 https://shaman.ceph.com/builds/ceph/wip-yuri-testing-2019-05-06-1633-mimic/

File: cmake/modules/Distutils.cmake [hunk 1]
```diff
@@ -63,7 +67,9 @@ function(distutils_install_cython_module name)
   install(CODE "
     set(ENV{CC} \"${PY_CC}\")
     set(ENV{LDSHARED} \"${PY_LDSHARED}\")
-    set(ENV{CPPFLAGS} \"-iquote${CMAKE_SOURCE_DIR}/src/include\")
+    set(ENV{CPPFLAGS} \"-iquote${CMAKE_SOURCE_DIR}/src/include
+                        -D'void0=dead_function\(void\)' \
```
Comment (tserong) [inline cmake/modules/Distutils.cmake]: Um. Guessing. Could the backslash at the end of this line be causing problems with string parsing? Or, conversely, the lack of backslash on the previous line?  That said, isn't it odd that it didn't fail when #27067 was tested?

**Comment by rgpublic** [comment]
Any possibility to speed this up? AFAIU there is only a pending reviewer (@rjfd), right? Currently we have a completely broken Ceph on Ubuntu 19.04 where no modules are loaded :-(
See also, for example: https://ruffell.nz/programming/writeups/2019/05/17/deploying-a-ceph-cluster-in-ubuntu-19.04.html

**Comment by smithfarm** [comment]
test this please

**Comment by edsokolowsk** [comment]
We are impacted by this issue.

Q1. What is the status of this branch release since we require ceph mimic running on Ubuntu 19.04? 

Q2. Is there any workaround , say using a previous release of Cython that supports subinterpreters?

**Comment by smithfarm** [comment]
retest this please

**Comment by smithfarm** [comment]
https://shaman.ceph.com/builds/ceph/wip-39593-mimic/8f8784f011b8df60b7646680aad0314f0dd656bc/

**Comment by smithfarm** [comment]
```
CMake Error at cmake/modules/Distutils.cmake:67 (install):
  Syntax error in cmake code at

    /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.2.6-91-g8f8784f/rpm/el7/BUILD/ceph-13.2.6-91-g8f8784f/cmake/modules/Distutils.cmake:67

  when parsing string

    
      set(ENV{CC} \"${PY_CC}\")
      set(ENV{LDSHARED} \"${PY_LDSHARED}\")
      set(ENV{CPPFLAGS} \"-iquote${CMAKE_SOURCE_DIR}/src/include
                          -D'void0=dead_function\(void\)' \
                          -D'__Pyx_check_single_interpreter\(ARG\)=ARG \#\# 0'\")
      set(ENV{LDFLAGS} \"-L${CMAKE_LIBRARY_OUTPUT_DIRECTORY}\")
      set(ENV{CYTHON_BUILD_DIR} \"${CMAKE_CURRENT_BINARY_DIR}\")
      set(ENV{CEPH_LIBDIR} \"${CMAKE_LIBRARY_OUTPUT_DIRECTORY}\")

  

      set(options --prefix=${CMAKE_INSTALL_PREFIX})
      if(DEFINED ENV{DESTDIR})
        if(EXISTS /etc/debian_version)
          list(APPEND options --install-layout=deb)
        endif()
        list(APPEND options --root=\$ENV{DESTDIR})
      else()
        list(APPEND options --root=/)
      endif()
      execute_process(
         COMMAND
             ${PYTHON${PYTHON_VERSION}_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/setup.py
             build --verbose --build-base ${CYTHON_MODULE_DIR}
             --build-platlib ${CYTHON_MODULE_DIR}/lib.${PYTHON${PYTHON_VERSION}_VERSION_MAJOR}
             build_ext --cython-c-in-temp --build-temp ${CMAKE_CURRENT_BINARY_DIR} --cython-include-dirs ${PROJECT_SOURCE_DIR}/src/pybind/rados
             install \${options} --single-version-externally-managed --record /dev/null
             egg_info --egg-base ${CMAKE_CURRENT_BINARY_DIR}
             --verbose
         WORKING_DIRECTORY \"${CMAKE_CURRENT_SOURCE_DIR}\"
         RESULT_VARIABLE install_res)
      if(NOT \"\${install_res}\" STREQUAL 0)
        message(FATAL_ERROR \"Failed to build and install ${name} python module\")
      endif()
    

  syntax error, unexpected cal_SYMBOL, expecting $end (252)
Call Stack (most recent call first):
  src/pybind/rados/CMakeLists.txt:3 (distutils_install_cython_module)


-- Configuring incomplete, errors occurred!
```

**Comment by smithfarm** [comment]
So, yeah, whichever version of CentOS that mimic is building on in Shaman is installing:

```
cmake-2.8.12.2-2.el7.x86_64
```

which does not understand the backslash syntax for line continuation.

File: cmake/modules/Distutils.cmake (hunk 1 above)
Comment (tchaikov) [reply cmake/modules/Distutils.cmake]: @tserong the reason why #27067 didn't fail the test is that we are using cmake3 since #22896.

**Comment by tchaikov** [comment]
as an alternative, we could backport #22896 .

**Comment by smithfarm** [comment]
https://shaman.ceph.com/builds/ceph/wip-39593-mimic/ce191414e28ebf247d77727219b634bc5507c34a/

**Comment by yuriw** [comment]
[wip-yuri2-testing-2019-07-11-1516-mimic](https://trello.com/c/DNupOYCk/664-wip-yuri2-testing-2019-07-11-1516-mimic)

**Pull request merged**